boofuzz/raw_l2_socket_connection.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import absolute_import
+
+import errno
+import socket
+import sys
+
+from future.utils import raise_
+
+from . import base_socket_connection, exception
+
+
+class RawL2SocketConnection(base_socket_connection.BaseSocketConnection):
+    """BaseSocketConnection implementation for use with Raw Layer 2 Sockets.
+
+    Args:
+        interface (str): Hostname or IP adress of target system.
+        send_timeout (float): Seconds to wait for send before timing out. Default 5.0.
+        recv_timeout (float): Seconds to wait for recv before timing out. Default 5.0.
+        ethernet_proto (int): Ethernet protocol to bind to. If supplied, the opened socket
+            gets bound to this protocol, otherwise the python default of 0 is used. Must
+            be supplied if this socket should be used for receiving. For valid options,
+            see <net/if_ether.h> in the Linux Kernel documentation. Usually, ETH_P_ALL
+            (0x0003) is not a good idea.
+        mtu (int): sets the maximum transmission unit size for this connection. Defaults
+            to 1518 for standard Ethernet.
```

Comment:
The Ethernet 2 frame consists of 14 bytes header, up to 1500 bytes data and 4 bytes CRC, totalling 1518 bytes.
However, setting this as the default value requires the NIC to not handle the CRC on it's own. While this might be possible with some NICs (check [this](https://stackoverflow.com/questions/723635/how-do-you-send-an-ethernet-frame-with-a-corrupt-fcs)), it's definitely not the default case for raw-l2 socket usage.
We should change the default value back to 1514.

Thoughts about this?

Reply:
That's why I introduced the has_framecheck argument (with a default value of True) to substract 4 bytes if the NIC calculates the CRC. Again, 802.11 cards usually don't inject the CRC; and for 802.5 the CRC is somewhere in the middle of the frame -- although I'm not sure about how much we can actually change in the 802.5 frame (and I don't currently have token ring equipment on hand to check, even if I'd be interested).

Reply:
My bad, I forgot about the has_framecheck argument.
Thanks for the quick reply, I'll finally merge this shortly.